src/Microsoft.Diagnostics.Monitoring/DiagnosticsEventPipeProcessor.cs

```diff
@@ -289,6 +317,96 @@ private void PostMetric(Metric metric)
             }
         }
 
+        private Task HandleGCEvents(EventPipeEventSource source, int pid, Func<Task> stopFunc, CancellationToken token)
+        {
+            int gcNum = -1;
+
+            TaskCompletionSource<object> gcEventsStarted = new TaskCompletionSource<object>();
+            token.Register(() => gcEventsStarted.TrySetCanceled(token));
+
+            TaskCompletionSource<object> gcStopped = new TaskCompletionSource<object>();
+            token.Register(() => gcStopped.TrySetCanceled(token));
+
+            source.Clr.GCStart += data =>
+            {
+                if (data.ProcessID != pid)
+                {
+                    return;
+                }
+
+                gcEventsStarted.TrySetResult(null);
+
+                if (gcNum < 0 && data.Depth == 2 && data.Type != GCType.BackgroundGC)
+                {
+                    gcNum = data.Count;
+                }
+            };
+
+            source.Clr.GCStop += data =>
+            {
+                if (data.ProcessID != pid)
+                {
+                    return;
+                }
+
+                if (data.Count == gcNum)
+                {
+                    gcStopped.TrySetResult(null);
+                }
+            };
+
+            source.Clr.GCBulkNode += data =>
+            {
+                if (data.ProcessID != pid)
+                {
+                    return;
+                }
+
+                gcEventsStarted.TrySetResult(null);
+            };
+
+            DotNetHeapDumpGraphReader dumper = new DotNetHeapDumpGraphReader(TextWriter.Null)
+            {
+                DotNetHeapInfo = new DotNetHeapInfo()
+            };
+            dumper.SetupCallbacks(_gcGraph, source, pid.ToString(CultureInfo.InvariantCulture));
+
+            return Task.Run(async () =>
+            {
+                // Wait for GC events for up to 5 seconds.
+                Task eventsStartedTask = gcEventsStarted.Task;
+                Task eventsTimeoutTask = Task.Delay(TimeSpan.FromSeconds(5), token);
+                Task completedTask = await Task.WhenAny(gcEventsStarted.Task, eventsTimeoutTask);
+
+                token.ThrowIfCancellationRequested();
+
+                if (completedTask == gcEventsStarted.Task)
+                {
+                    // If started receiving GC events, wait for the GC Stop event and then
+                    // stop processing events.
+                    await gcStopped.Task;
+
+                    token.ThrowIfCancellationRequested();
+
+                    await stopFunc();
+                }
+                else
+                {
+                    // If have not received any GC events yet, stop processing events, which will
+                    // force flushing of the GC events and then wait for the GC Stop event.
+                    await stopFunc();
```

Comment:
we're awaiting stopped without first awaiting started here. Is it possible that the gc stopped event never happens?

Reply:
This is the reason for the type flush: https://github.com/microsoft/perfview/blob/6e9043ace41ed0917cf566f387f4a7cce5e26a57/src/EtwHeapDump/DotNetHeapDumper.cs#L35-L38